v2/reuse_test.go

```diff
@@ -0,0 +1,578 @@
+package reuseport
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"os/exec"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+)
+
+func ResolveAddr(network, address string) (net.Addr, error) {
+	switch network {
+	default:
+		return nil, net.UnknownNetworkError(network)
+	case "ip", "ip4", "ip6":
+		return net.ResolveIPAddr(network, address)
+	case "tcp", "tcp4", "tcp6":
+		return net.ResolveTCPAddr(network, address)
+	case "udp", "udp4", "udp6":
+		return net.ResolveUDPAddr(network, address)
+	case "unix", "unixgram", "unixpacket":
+		return net.ResolveUnixAddr(network, address)
+	}
+}
+
+func Listen(network, addr string) (net.Listener, error) {
+	l := net.ListenConfig{
+		Control: Control,
+	}
+	return l.Listen(context.Background(), network, addr)
+}
+func ListenPacket(network, addr string) (net.PacketConn, error) {
+	l := net.ListenConfig{
+		Control: Control,
+	}
+	return l.ListenPacket(context.Background(), network, addr)
+}
+
+func Dial(network, laddr, raddr string) (net.Conn, error) {
+	var d net.Dialer
+	if laddr != "" {
+		netladdr, err := ResolveAddr(network, laddr)
+		if err != nil {
+			return nil, err
+		}
+		d.LocalAddr = netladdr
+	}
+	d.Control = Control
+	return d.Dial(network, raddr)
+}
+
+type readOnly struct {
+	r io.Reader
+}
+
+func (r readOnly) Read(buf []byte) (n int, err error) {
+	return r.r.Read(buf)
+}
+
+func echo(c net.Conn) {
+	io.Copy(c, readOnly{c})
```

Comment:
This is to remove optimization added in Go 1.11 which uses splice but also doubles required file descriptors.

The optimization is only used when copying from TCP to TCP.